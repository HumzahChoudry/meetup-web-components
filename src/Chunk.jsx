import React from 'react';
import cx from 'classnames';

export const CHUNK_CLASS = 'chunk';
/**
 * Design System Component: Provides `stripe` styled container for components
 * @module Chunk
 */
class Chunk extends React.Component {
	render() {
		const {
			children,
			className,
			...other
		} = this.props;

		const classNames = cx(
			CHUNK_CLASS,
			className
		);

		return (
			<div
				className={classNames}
				{...other}
			>
				{children}
			</div>
		);
	}
}

export default Chunk;
