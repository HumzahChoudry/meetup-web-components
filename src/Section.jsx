import React from 'react';
import cx from 'classnames';

/**
 * @module Section
 */
class Section extends React.Component {
	render() {
		const {
			children,
			className,
			noBorder,
			...other
		} = this.props;

		const classNames = cx(
			'section',
			{
				noBorder: noBorder
			},
			className
		);

		return (
			<div
				className={classNames}
				{...other}>
					{children}
			</div>
		);
	}
}

Section.propTypes = {
	noBorder: React.PropTypes.bool
};

export default Section;
