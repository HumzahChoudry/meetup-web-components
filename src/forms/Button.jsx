import React from 'react';
import cx from 'classnames';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

export const BUTTON_CLASS = 'button';
export const BUTTON_ICON_WRAPPER_CLASS = 'button-icon-wrapper';
export const BUTTON_LABEL_CLASS = 'button-label';
export const BUTTON_ICON_CLASS = 'button-icon';

/**
 * SQ2 Button component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_buttons.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#buttons}
 * @module Button
 */
class Button extends React.Component {

	render() {
		const {
			children,
			className,
			onClick,
			reset,
			fullWidth,
			icon,
			primary,
			right,
			small,
			...other
		} = this.props;

		const classNames = cx(
			BUTTON_CLASS,
			{
				'button--fullWidth': fullWidth,
				'button--primary': primary,
				'button--small': small,
				'button--reset': reset,
			},
			className
		);
		const opts = right ? { rowReverse: 'all' } : {};

		const iconChildren = (
			<Flex className={BUTTON_ICON_WRAPPER_CLASS} {...opts}>
				{icon &&
					<FlexItem shrink className={`${BUTTON_ICON_CLASS} valign--middle`}>
						{icon}
					</FlexItem>
				}
				{children &&
					<FlexItem className={`${BUTTON_LABEL_CLASS} valign--middle align--center atMedium_align--left`}>
						{children}
					</FlexItem>
				}
			</Flex>
		);

		return (
			<button
				className={classNames}
				onClick={onClick}
				role='button'
				{...other}
			>
				{ icon ? iconChildren : children }
			</button>
		);
	}
}

Button.defaultProps = {
	right: false
};

Button.propTypes = {
	reset: React.PropTypes.bool,
	fullWidth: React.PropTypes.bool,
	primary: React.PropTypes.bool,
	small: React.PropTypes.bool,
	icon: React.PropTypes.any,
	right: React.PropTypes.bool,
};
export default Button;
