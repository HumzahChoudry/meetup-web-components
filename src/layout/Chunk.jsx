// @flow

import React from 'react';
import cx from 'classnames';

import withLoading from '../utils/components/withLoading';

type Props = {|
	/** The child elements of the component */
	children: React$Node,
	className?: string,
	/** Whether the component is in a loading state */
	isLoading?: boolean,
	/** Props to pass to the `<Loading />` component */
	loadingProps?: {
		color?: string,
		scrimColor?: string,
		size?: MediaSizes,
	},
|};
/**
 * Design System Component: Provides `stripe` styled container for components
 * @module ChunkComponent
 */
export class ChunkComponent extends React.Component<Props> {
	render() {
		const {
			children,
			className,
			loadingProps = {}, // eslint-disable-line no-unused-vars
			isLoading,
			...other
		} = this.props;

		const classNames = cx('chunk', { 'component--isLoading': isLoading }, className);

		return (
			<div className={classNames} {...other}>
				{children}
			</div>
		);
	}
}

const Chunk = withLoading(ChunkComponent);
Chunk.displayName = 'Chunk';
export default Chunk;
