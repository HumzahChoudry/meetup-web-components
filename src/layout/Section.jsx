import React from 'react';
import cx from 'classnames';

export const SECTION_CLASS = 'section';
export const SECTION_NOBORDER_CLASS = 'section--noBorder';
export const VALID_BREAKPOINTS = {
	all: 'atAll',
	medium: 'atMedium',
	large: 'atLarge',
};

/**
 * Design System Component: Provides `section` container for components
 * @module Section
 */
class Section extends React.Component {
	render() {
		const {
			children,
			className,
			noBorder,
			...other
		} = this.props;

		const noBorderBreakpoint = VALID_BREAKPOINTS[noBorder] || VALID_BREAKPOINTS['all'];

		const classNames = cx(
			SECTION_CLASS,
			{
				[`${noBorderBreakpoint}_${SECTION_NOBORDER_CLASS}`]: noBorder,
			},
			className
		);

		return (
			<section
				className={classNames}
				{...other}
			>
				{children}
			</section>
		);
	}
}
Section.propTypes = {
	noBorder: React.PropTypes.oneOf(Object.keys(VALID_BREAKPOINTS))
};

export default Section;
