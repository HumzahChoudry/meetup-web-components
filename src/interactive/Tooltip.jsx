import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import Button from '../forms/Button';
import Icon from '../media/Icon';
import FloatingPosition from '../utils/components/FloatingPosition';

/**
 * @module Tooltip
 */
class Tooltip extends React.PureComponent {
	constructor(props) {
		super(props);

		this.onBlur = this.onBlur.bind(this);
		this.closeContent = this.closeContent.bind(this);
		this.openContent = this.openContent.bind(this);

		this.state = {
			isActive: props.isActive || false,
		};
	}

	closeContent(e) {
		this.setState(() => ({ isActive: false }));

		if (this.props.onBlur) {
			this.props.onBlur(e);
		}
		if (this.props.onMouseLeave) {
			this.props.onMouseLeave(e);
		}
	}

	openContent(e) {
		this.setState(() => ({ isActive: true }));

		if (this.props.onFocus) {
			this.props.onFocus(e);
		}
		if (this.props.onMouseEnter) {
			this.props.onMouseEnter(e);
		}
	}

	onBlur(e) {
		if (!this.contentRef) return;

		setTimeout(() => {
			if (!this.contentRef.contains(document.activeElement)) {
				this.closeContent(e);
			}
		}, 1);
	}

	render() {
		const {
			className,
			trigger,
			content,
			align,
			offset,
			maxWidth,
			minWidth,
			noPortal,
			id,
			direction,
			manualToggle,
			withClose,
			...other
		} = this.props;

		// Do not pass along to children
		delete other.isActive;

		const classNames = {
			dropdown: cx(className, 'popup', {
				'popup--noPortal': noPortal,
			}),
		};

		const isActive = manualToggle ? this.props.isActive : this.state.isActive;

		const getTrigger = () => {
			return this.triggerRef;
		};

		const getContent = () => {
			return this.contentRef;
		};

		return (
			<div
				className={classNames.dropdown}
				onMouseLeave={this.closeContent}
				{...other}
			>
				<div
					ref={el => (this.triggerRef = el)}
					className="popup-trigger"
					onFocus={this.openContent}
					onBlur={this.onBlur}
					onMouseEnter={this.openContent}
					aria-labelledby={id}
				>
					{trigger}
				</div>
				{isActive && (
					<FloatingPosition
						getTrigger={getTrigger}
						getContent={getContent}
						noPortal={noPortal}
						align={align}
						offset={offset}
						direction={direction}
					>
						{({ top, left, align, boundedMaxWidth }) => (
							<div
								ref={el => (this.contentRef = el)}
								className={cx('popup-content tooltip-content', {
									'popup-content--right': align === 'right',
									'popup-content--left': align === 'left',
									'popup-content--center': align === 'center',
									'popup-content--top': direction === 'top',
									'display--none': !isActive,
									'display--block': isActive,
								})}
								aria-hidden={!isActive}
								id={id}
								role="tooltip"
								style={{
									left: left,
									top: top,
									minWidth: minWidth,
									maxWidth: maxWidth || boundedMaxWidth,
								}}
							>
								<div
									className={cx('popup-bubble tooltip-bubble inverted', {
										'popup-bubble--active': isActive,
										'popup-bubble--right': align === 'right',
										'popup-bubble--left': align === 'left',
										'popup-bubble--center': align === 'center',
										'popup-bubble--top': direction === 'top',
									})}
								>
									{withClose && (
										<Button
											className="tooltip-closeBtn"
											onClick={this.closeContent}
											icon={<Icon shape="cross" className="text--secondary" />}
											reset
										/>
									)}
									{content}
								</div>
							</div>
						)}
					</FloatingPosition>
				)}
			</div>
		);
	}
}

Tooltip.defaultProps = {
	direction: 'bottom',
	minWidth: '0px',
	noPortal: false,
};

Tooltip.propTypes = {
	id: PropTypes.string.isRequired,
	trigger: PropTypes.element.isRequired,
	content: PropTypes.element,
	align: PropTypes.oneOf(['left', 'right', 'center']).isRequired,
	offset: PropTypes.shape({
		left: PropTypes.number,
		top: PropTypes.number,
	}),
	direction: PropTypes.oneOf(['top', 'bottom']).isRequired,
	className: PropTypes.string,
	isActive: PropTypes.bool,
	manualToggle: PropTypes.bool,
	maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	noPortal: PropTypes.bool,
};

export default Tooltip;
