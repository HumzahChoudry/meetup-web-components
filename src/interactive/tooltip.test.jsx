import React from 'react';
import { shallow, mount } from 'enzyme';
import Section from '../layout/Section';
import Chunk from '../layout/Chunk';
import Button from '../forms/Button';

import Tooltip from './Tooltip';

const dropdownContent = (
	<Section noSeparator>
		<Chunk>
			<h2 className="text--big text--bold">Dropdown content</h2>
		</Chunk>
	</Section>
);
const dropdownTrigger = <Button small>Open</Button>;

describe('Tooltip', () => {
	const dropdownJSX = (
		<Tooltip
			align="right"
			trigger={dropdownTrigger}
			content={dropdownContent}
		/>
	);
	const wrapper = shallow(dropdownJSX);

	it('should hide tooltip content by default', () => {
		expect(wrapper.find('.dropdown-content').length).toBeFalsy();
	});

	describe('right aligned tooltip', () => {
		const rightDropdown = (
			<Tooltip
				align="right"
				trigger={dropdownTrigger}
				content={dropdownContent}
			/>
		);
		const rightDropdownWrapper = mount(rightDropdown);
		const trigger = rightDropdownWrapper.find('.dropdown-trigger');

		it('applies correct alignment className to tooltip content', () => {
			trigger.simulate('focus');
			const content = rightDropdownWrapper.find('.dropdown-content');

			expect(content.prop('className')).toContain('dropdown-content--right');
		});
	});

	describe('right aligned tooltip', () => {
		const aboveDropdown = (
			<Tooltip
				popAbove
				trigger={dropdownTrigger}
				content={dropdownContent}
			/>
		);
		const aboveDropdownWrapper = mount(aboveDropdown);
		const trigger = aboveDropdownWrapper.find('.dropdown-trigger');

		it('applies correct alignment className to tooltip content', () => {
			trigger.simulate('focus');
			const content = aboveDropdownWrapper.find('.dropdown-content');

			expect(content.prop('className')).toContain('dropdown-content--above');
		});
	});

	describe('open and close', () => {
		let closedComponent, trigger, tooltipArea;

		beforeEach(() => {
			closedComponent = shallow(dropdownJSX);
			trigger = closedComponent.find('.dropdown-trigger').first();
			tooltipArea = closedComponent.find('.dropdown').first();
		});
		afterEach(() => {
			closedComponent = null;
			trigger = null;
			tooltipArea = null;
		});

		it('should show tooltip when trigger is hovered', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('mouseEnter');
			expect(closedComponent.state('isActive')).toBeTruthy();
		});

		it('should hide tooltip when cursor leaves the tooltip area', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('mouseEnter');
			expect(closedComponent.state('isActive')).toBeTruthy();
			tooltipArea.simulate('mouseLeave');
			expect(closedComponent.state('isActive')).toBeFalsy();
		});

		it('should show tooltip when trigger is focused', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('focus');
			expect(closedComponent.state('isActive')).toBeTruthy();
		});

		it('should hide tooltip when trigger is blurred', () => {
			expect(closedComponent.state('isActive')).toBeFalsy();
			trigger.simulate('focus');
			expect(closedComponent.state('isActive')).toBeTruthy();
			tooltipArea.simulate('blur');

			setTimeout(() => {
				expect(closedComponent.state('isActive')).toBeFalsy();
			}, 10);
		});
	});
});
