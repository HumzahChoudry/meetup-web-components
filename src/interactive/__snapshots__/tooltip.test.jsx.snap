// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip aligned tooltip matches snapshot 1`] = `
<Tooltip
  align="right"
  content={
    <Section
      noSeparator={true}
    >
      <Chunk>
        <h2
          className="text--big text--bold"
        >
          Dropdown content
        </h2>
      </Chunk>
    </Section>
  }
  direction="top"
  id="testDropdown"
  maxWidth="384px"
  minWidth="0px"
  noPortal={false}
  trigger={
    <Button
      component="button"
      small={true}
      type="button"
    >
      Open
    </Button>
  }
/>
`;

exports[`Tooltip tooltip with a close button should render correctly 1`] = `
<Tooltip
  align="right"
  content={
    <Section
      noSeparator={true}
    >
      <Chunk>
        <h2
          className="text--big text--bold"
        >
          Dropdown content
        </h2>
      </Chunk>
    </Section>
  }
  direction="bottom"
  id="testDropdown"
  isActive={true}
  maxWidth="384px"
  minWidth="0px"
  noPortal={false}
  trigger={
    <Button
      component="button"
      small={true}
      type="button"
    >
      Open
    </Button>
  }
  withClose={true}
>
  <div
    className="popup"
    onMouseLeave={[Function]}
  >
    <div
      aria-labelledby="testDropdown"
      className="popup-trigger"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseEnter={[Function]}
    >
      <Button
        component="button"
        small={true}
        type="button"
      >
        <button
          className="button button--small"
          type="button"
        >
          Open
        </button>
      </Button>
    </div>
    <FloatingPosition
      align="right"
      direction="bottom"
      getContent={[Function]}
      getTrigger={[Function]}
      noPortal={false}
      offset={undefined}
    >
      <ConditionalWrap
        condition={true}
        wrap={[Function]}
      >
        <Portal>
          <div
            aria-hidden={false}
            className="popup-content tooltip-content popup-content--right display--block"
            id="testDropdown"
            role="tooltip"
            style={
              Object {
                "left": "0px",
                "maxWidth": "384px",
                "minWidth": "0px",
                "top": "0px",
              }
            }
          >
            <div
              className="popup-bubble tooltip-bubble inverted popup-bubble--active popup-bubble--right"
            >
              <Button
                className="tooltip-closeBtn"
                component="button"
                icon={
                  <Icon
                    className="text--secondary"
                    shape="cross"
                    size="s"
                  />
                }
                onClick={[Function]}
                reset={true}
                type="button"
              >
                <button
                  className="button button--reset button--iconOnly tooltip-closeBtn"
                  onClick={[Function]}
                  type="button"
                >
                  <Flex
                    className="button-icon-wrapper"
                    justify="center"
                  >
                    <Flex
                      aria-busy={undefined}
                      className="button-icon-wrapper"
                      direction="row"
                      isLoading={undefined}
                      justify="center"
                    >
                      <div
                        aria-busy={undefined}
                        className="flex flex--row flex--center button-icon-wrapper"
                      >
                        <FlexItem
                          className="button-icon valign--middle flush--left"
                          shrink={true}
                        >
                          <FlexItem
                            aria-busy={undefined}
                            className="button-icon valign--middle flush--left"
                            isLoading={undefined}
                            shrink={true}
                          >
                            <div
                              aria-busy={undefined}
                              className="flex-item flex-item--shrink button-icon valign--middle flush--left"
                            >
                              <Icon
                                className="text--secondary"
                                shape="cross"
                                size="s"
                              >
                                <span
                                  className="svg svg--cross text--secondary"
                                >
                                  <svg
                                    className="svg-icon valign--middle"
                                    height="24"
                                    preserveAspectRatio="xMinYMin meet"
                                    role="img"
                                    style={Object {}}
                                    viewBox="0 0 24 24"
                                    width="24"
                                  >
                                    <use
                                      xlinkHref="#icon-cross--small"
                                    />
                                  </svg>
                                </span>
                              </Icon>
                            </div>
                          </FlexItem>
                        </FlexItem>
                      </div>
                    </Flex>
                  </Flex>
                </button>
              </Button>
              <Section
                noSeparator={true}
              >
                <Section
                  aria-busy={undefined}
                  isLoading={undefined}
                  noSeparator={true}
                >
                  <section
                    aria-busy={undefined}
                    className="section"
                  >
                    <Chunk>
                      <Chunk
                        aria-busy={undefined}
                        isLoading={undefined}
                      >
                        <div
                          aria-busy={undefined}
                          className="chunk"
                        >
                          <h2
                            className="text--big text--bold"
                          >
                            Dropdown content
                          </h2>
                        </div>
                      </Chunk>
                    </Chunk>
                  </section>
                </Section>
              </Section>
            </div>
          </div>
        </Portal>
      </ConditionalWrap>
    </FloatingPosition>
  </div>
</Tooltip>
`;
